From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Revanth Kumar Uppala <ruppala@nvidia.com>
Date: Mon, 2 Mar 2026 12:04:15 +0000
Subject: [PATCH 233/243] NVIDIA: SAUCE: PCI: Add support for PEX WAKE GPIO

Add support for handling the PEX WAKE GPIO (pex_wake_gpiod) by
parsing the pex-wake GPIO from the device tree and setup the
interrupt line.

Enable and disable the wake-up interrupt during system
suspend/resume.

Bug 4701216

Change-Id: I34aabb447d72e2f173fbeebcb8cee09edd211693
Signed-off-by: ruppala <ruppala@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-jammy/+/3210115
(cherry picked from commit 34f394b80575b555eabc7395f767344e2adb571c)
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3562156
Reviewed-by: mobile promotions <svcmobile_promotions@nvidia.com>
Reviewed-by: Brad Griffis <bgriffis@nvidia.com>
(cherry picked from commit 0346e0b0f245e618d11b4d37ebb846ded2c22ab9)
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3582750
Reviewed-by: Prathamesh Shete <pshete@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
APS-Inflight: svcenomis <svcenomis@nvidia.com>
Tested-by: Prathamesh Shete <pshete@nvidia.com>
Reviewed-by: Jon Hunter <jonathanh@nvidia.com>
(cherry picked from commit cd3994cc8533ed624016c837cad155cfabb2f84c)

diff --git a/drivers/pci/controller/dwc/pcie-tegra194.c b/drivers/pci/controller/dwc/pcie-tegra194.c
index 23db64b402557..7f6641d747a25 100644
--- a/drivers/pci/controller/dwc/pcie-tegra194.c
+++ b/drivers/pci/controller/dwc/pcie-tegra194.c
@@ -291,6 +291,8 @@ struct tegra_pcie_dw {
 	struct gpio_desc *pex_rst_gpiod;
 	struct gpio_desc *pex_refclk_sel_gpiod;
 	struct gpio_desc *pex_prsnt_gpiod;
+	struct gpio_desc *pex_wake_gpiod;
+	unsigned int pex_wake_irq;
 	unsigned int pex_rst_irq;
 	unsigned int prsnt_irq;
 	bool perst_irq_enabled;
@@ -1171,6 +1173,29 @@ static int tegra_pcie_dw_parse_dt(struct tegra_pcie_dw *pcie)
 		return ret;
 	}
 
+	pcie->pex_wake_gpiod = devm_gpiod_get_optional(pcie->dev, "nvidia,pex-wake",
+			(pcie->of_data->mode == DW_PCIE_RC_TYPE) ? GPIOD_IN : GPIOD_OUT_HIGH);
+	if (IS_ERR(pcie->pex_wake_gpiod)) {
+		int err = PTR_ERR(pcie->pex_wake_gpiod);
+
+		if (err == -EPROBE_DEFER) {
+			dev_info(pcie->dev, "pex-wake GPIO probe deferred\n");
+			return err;
+		}
+		pcie->pex_wake_gpiod = NULL;
+	}
+
+	if (!IS_ERR_OR_NULL(pcie->pex_wake_gpiod)) {
+		device_init_wakeup(pcie->dev, true);
+
+		ret = gpiod_to_irq(pcie->pex_wake_gpiod);
+		if (ret < 0) {
+			dev_err(pcie->dev, "Failed to get IRQ for WAKE GPIO: %d\n", ret);
+			return ret;
+		}
+		pcie->pex_wake_irq = (unsigned int)ret;
+	}
+
 	ret = of_property_count_strings(np, "phy-names");
 	if (ret < 0) {
 		dev_err(pcie->dev, "Failed to find PHY entries: %d\n",
@@ -2554,6 +2579,7 @@ static int tegra_pcie_dw_suspend_late(struct device *dev)
 static int tegra_pcie_dw_suspend_noirq(struct device *dev)
 {
 	struct tegra_pcie_dw *pcie = dev_get_drvdata(dev);
+	int ret = 0;
 
 	if (pcie->of_data->mode == DW_PCIE_EP_TYPE)
 		return 0;
@@ -2565,6 +2591,11 @@ static int tegra_pcie_dw_suspend_noirq(struct device *dev)
 	tegra_pcie_dw_pme_turnoff(pcie);
 	tegra_pcie_unconfig_controller(pcie);
 
+	if (pcie->pex_wake_gpiod && device_may_wakeup(dev)) {
+		ret = enable_irq_wake(pcie->pex_wake_irq);
+		if (ret < 0)
+			dev_err(dev, "enable wake irq failed: %d\n", ret);
+	}
 	return 0;
 }
 
@@ -2579,6 +2610,12 @@ static int tegra_pcie_dw_resume_noirq(struct device *dev)
 	if (!pcie->link_state)
 		return 0;
 
+	if (pcie->pex_wake_gpiod && device_may_wakeup(dev)) {
+		ret = disable_irq_wake(pcie->pex_wake_irq);
+		if (ret < 0)
+			dev_err(dev, "disable wake irq failed: %d\n", ret);
+	}
+
 	ret = tegra_pcie_config_controller(pcie, true);
 	if (ret < 0)
 		return ret;
-- 
2.43.0

